### PR TITLE
Personnalisation

### PR DESCRIPTION
**Pull request by HugoSenetaire**
Add different method for personnalisation of training.
Add more natural save and load models (save parameters but also personnalisation state)

### CODE REVIEW

File: agentLocal.py [hunk 1]
```diff
@@ -45,28 +45,27 @@ def get_action(self, obs):
 
 
 class DFPAgent():
-    def __init__(self, state_size, measurement_names, action_size, timesteps, use_cuda=False):
-      self.measurement_names = measurement_names
-      # get size of state, measurement, action, and timestep
-      self.state_size = state_size #sensory input size
+    def __init__(self, state_size, measurement_names, action_size, timesteps, args, use_cuda=False):
```
Comment (MatthieuMoreau0) [inline agentLocal.py]: les arguments du constructeur ne sont plus necessaires

File: default_parameters.py [hunk 2]
```diff
@@ -1,49 +1,93 @@
 import argparse
 import torch
+import datetime
+import os
 
 
 def create_args():
     parser = argparse.ArgumentParser(description='DFP exp')
     parser.add_argument('--gpu', action='store_true', help='Whether to use GPU')
+    parser.add_argument('--experiment_name', type=str, default = "None")
```
Comment (MatthieuMoreau0) [inline default_parameters.py]: None en string ?

File: default_parameters.py [hunk 3]
```diff
@@ -1,49 +1,93 @@
...
+    parser.add_argument('--GLOBAL_PATH', type=str, default="experiments/")
```
Comment (MatthieuMoreau0) [inline default_parameters.py]: bizarre que cet arg soit en maj contrairement aux autres

File: default_parameters.py [hunk 4]
```diff
@@ -1,49 +1,93 @@
...
+        "running_in_notebook" : True,
+        "DEBUG": False,
+        "SAVE_VIDEO" : False,
+    })
+
+
+
     # Games update :
     args.update({
+        "SCENARIO" : "11_vs_11_kaggle",
```
Comment (MatthieuMoreau0) [inline default_parameters.py:50]: peut etre ajouter le scenar dans les args du script

File: default_parameters.py [hunk 5]
```diff
@@ -1,49 +1,93 @@
...
         "CHANNEL_NAMES" : ["left_team", 'right_team', 'left_team_direction', 'right_team_direction', 'ball', 'ball_direction'],
         "MEASUREMENT_NAMES" : ['goals', 'ball_distance_to_goal', "ball_distance_to_center", "possession"],
         "TIMESTEPS" : [1,2,4,8,16,32],
         "IMAGE_SIZE" : (43, 101),
         "NB_ACTIONS" : 19,
+        "NUM_STEPS" : 200
```
Comment (MatthieuMoreau0) [inline default_parameters.py]: idem ajouter dans les args ? (au mois avec default value)

File: default_parameters.py [hunk 6]
```diff
@@ -1,49 +1,93 @@
...
     })
 
+        # Save update
+
+    args["save_every"] = 5000
+
+    if args["GLOBAL_PATH"] == "experiments/":
```
Comment (MatthieuMoreau0) [inline default_parameters.py]: je trouve ca un peu bizarre comme condition perso, c'est pas tres intuitif à l'usage meme si je comprends bien que c'est le truc par défaut

File: default_parameters.py [hunk 7]
```diff
@@ -1,49 +1,93 @@
...
+        if args["running_in_notebook"]:
+            args["GLOBAL_PATH"] =  "/content/drive/My Drive/google-football/"
```
Comment (MatthieuMoreau0) [inline default_parameters.py]: mais je trouve ca cool comme principe de faire ca automatiquement pour les notebook

d'ailleurs l'argument `running_in_notebook` est aussi utilisé dans la création de l'environnement football, je me demande ce que ca change (il faudra peut etre faire deux arguments différents en fonction de a quoi ca sert)

File: default_parameters.py [hunk 8]
```diff
@@ -1,49 +1,93 @@
...
+    
+    if args["experiment_name"]=="None" :
+        currentdate = str(datetime.datetime.now())
+        args["experiment_name"] = args["SCENARIO"] + "_" + currentdate  
+        
+    total_path = os.path.join(args["GLOBAL_PATH"],args["experiment_name"])
+    if not os.path.exists(total_path):
+        os.makedirs(total_path)
+
+    args["TOTAL_PATH"] = total_path
```
Comment (MatthieuMoreau0) [inline default_parameters.py:66]: toute la logique de création de `TOTAL_PATH` je le mettrais dans une fonction dédiée get_total_path pour que cette fonction `update_default_args` soit pas trop longue

File: default_parameters.py [hunk 9]
```diff
@@ -1,49 +1,93 @@
...
+        "final_epsilon": 0.0001,
         "batch_size" : 32,
         "explore" : 100000,
         "observe" : 2000,
         "frame_per_action" : 4,
         "timestep_per_train" : 5,
         "max_memory" : 20000,
-        "NUM_STEPS" : 200
+
     })
```
Comment (MatthieuMoreau0) [inline default_parameters.py:83]: est ce qu'on ferait pas des sous dictionnaires d'arguments, genre les arguments de l'environnements d'un côté, les arguments de l'agent d'un autre, les arguments de log/save, model, scheduler ...

Si on save la config de chaque expérience ca sera plus lisible

File: default_parameters.py [hunk 10]
```diff
@@ -1,49 +1,93 @@
...
 
 
+    args["total_train"] = 2000000
+    assert(args["total_train"]>=args["explore"]+args["observe"])
```
Comment (MatthieuMoreau0) [inline default_parameters.py:87]: il se passe quoi pour t>= explore + observe ?

File: run_exp.py [hunk 11]
```diff
@@ -6,14 +6,46 @@
...
+import json
+import pickle
 
 
 if __name__ == '__main__':
     args = create_args()
     update_default_args(args)
+    if args["load_model"]:
+        print("Load parameter from {}".format(os.path.join(args["TOTAL_PATH"],'argument.json')))
+        t = args["t"]
```
Comment (MatthieuMoreau0) [inline run_exp.py]: pourquoi on change la valeur de t puis celle de `args["t"]`

File: utils.py [hunk 12]
```diff
@@ -118,3 +127,49 @@ def create_goal(rel_goal, timesteps_size):
...
+    return True
+  
+def find_last_iteration(args):
+  total_path = args["TOTAL_PATH"]
+  list_trained = glob.glob(os.path.join(total_path,"*.pth"))
+  if not os.path.exists or len(list_trained) == 0 :
+    raise NameError("No experiments previously trained at this path")
+  list_t = list(map(lambda x: int(x.strip(os.path.join(total_path, "weights_")).strip(".pth")),list_trained)) # TODO : Changer ce monstre
+  list_t.sort()
+  args["t"] = list_t[-1]
```
Comment (MatthieuMoreau0) [inline utils.py]: return plutot que changer en soum-soum

File: agentLocal.py (hunk 1 above)
Comment (HugoSenetaire) [reply agentLocal.py]: Already changed

File: default_parameters.py (hunk 2 above)
Comment (HugoSenetaire) [reply default_parameters.py]: A essayer sans string

File: default_parameters.py (hunk 3 above)
Comment (HugoSenetaire) [reply default_parameters.py]: Frère c'était la méta

File: default_parameters.py (hunk 4 above)
Comment (HugoSenetaire) [reply default_parameters.py:50]: Oui

File: default_parameters.py (hunk 5 above)
Comment (HugoSenetaire) [reply default_parameters.py]: Oui

File: default_parameters.py (hunk 8 above)
Comment (HugoSenetaire) [reply default_parameters.py:66]: Ok

File: default_parameters.py (hunk 9 above)
Comment (HugoSenetaire) [reply default_parameters.py:83]: Oui

File: default_parameters.py (hunk 10 above)
Comment (HugoSenetaire) [reply default_parameters.py:87]: vu ensemble

File: utils.py (hunk 12 above)
Comment (HugoSenetaire) [reply utils.py]: OK

**Comment by MatthieuMoreau0** [comment]
Most changes have been added, needs to refactor arguments into sub dictionnaries

File: default_parameters.py [hunk 13]
```diff
@@ -8,14 +8,31 @@ def create_args():
...
 
 
+def get_total_path(args):
+    if args["global_path"] is None:
+        if args["running_in_notebook"]:
+            args["global_path"] =  "/content/drive/My Drive/google-football/"
+        else:
+            args["global_path"] = "experiments/"
+    
+    if args["experiment_name"]=="None" :
```
Comment (HugoSenetaire) [inline default_parameters.py]: Do same here
